docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,372 @@
+* Use exceptions *only* for exceptional circumstances.
+* Handle normal errors gracefully.
+
+* Do not rely on exceptions to avoid taking responsibility.
+* Do not throw exceptions in destructors as this can crash the program.
+
+
+## Memory handling
+
+* Use references where the value cannot be null. Use pointers where it can.
+* Make sure no uninitialised memory is accessed.
+* Always prefer passing by reference or `const` reference where possible.
+* Prefer smart pointers (`std::shared_ptr`, `std::unique_ptr`, etc) over raw pointers.
+* Use `make_shared` and `make_unique` to wrap objects in smart pointers.
+* Prefer the type-safe `nullptr` over the weakly-typed literal `0` and the `NULL` macro.
+* Stick to RAII principle and automatic memory management - destructors are ideally empty.
+* When working with array object types, prefer existing structures that already implement memory management:
+  * From the standard library, such as `std::vector` and `std::string`.
+  * `SharedArray` and `Array` classes from `fetch::memory` namespace.
+  * Classes from `fetch::byte_array` namespace, such as `BasicByteArray`.
+* It is safe to transfer ownership from `std::unique_ptr` to `std::shared_ptr`. However, do **NOT** transfer ownership the other way around, as that will result in memory access exceptions; in this case a non-owning `std::weak_ptr` should be used instead of `std::unique_ptr` (see [reference documentation](https://en.cppreference.com/w/cpp/memory/weak_ptr)).
+
+Example of using smart pointer with **exclusive** ownership:
+
+``` c++
+#include <memory>
+
+// Parent Scope block
+{
+  std::unique_ptr<int> a0;
+
+  // Nested scope block
+  {
+    std::unique_ptr<int> a1 = std::make_unique<int>(5);
+    a0                      = a1;  // Transfer of OWNERSHIP from `a1` to `a0`
+    // `a1` does NOT control lifecycle of the `int` instance anymore.
+
+    // This will result in throwing an exception since `a1` is now `nullptr`
+    std::cout << "a1=" << *a1 << std::endl;
+
+    // The automatically called destructor of `a1` does NOTHING since
+    // `a1` does NOT control any object.
+  }
+
+  // Automatically called destructor of `a0` destroys the `int` instance
+  // allocated and initialised in the nested scope above.
+}
+```
+
+
+## Resource handling
+
+* Use a smart pointer to manage resources whose lifecycle needs strict control, such as:
+  * Thread synchronisation constructs (e.g. mutex).
+  * Filesystem objects.
+  * Network connections.
+  * Database sessions.
+
+* A destructor must **NEVER** throw an exception. The reason for this is that when the destructor is called automatically because an exception is thrown, if said destructor throws another one, the C++ runtime will call `std::terminate` of the process (as per the [C++ standard definition](http://en.cppreference.com/w/cpp/language/destructor#Exceptions)).
+
+* Use locking guards provided in C++ `std` namespace, such as `std::unique_lock` and `std::lock_guard`.
```

Comment:
We have a macro for locking that appears to be favoured by @ejfitzgerald - FETCH_LOCK - might be worth saying to use this